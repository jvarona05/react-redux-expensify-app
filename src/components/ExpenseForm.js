import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates'

export default class ExpenseForm extends React.Component {
    constructor(props){
        super(props)

        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? props.expense.amount : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        }
    }
    
    onDescriptionChange = (e) => {
        const description = e.target.value
        this.setState(() => ({description}))
    }
    onNoteChange = (e) => {
        const note = e.target.value
        this.setState(() => ({note}))
    }
    onAmountChange = (e) => {
        const amount = e.target.value
        if(!amount || amount.match(/^[0-9]+(\.[0-9]{1,2})?$/)) {
            this.setState(() => ({amount}))
        }
    }
    onSubmit = (e) => {
        e.preventDefault()
        
        if(!this.state.description || !this.state.amount) {
            this.setState(() => ({error: 'Please provide description and amount'}))
        } else {
            this.setState(() => ({error: ''}))
            this.props.onSubmit({
                description: this.state.description,
                amount: this.state.amount,
                note: this.state.note,
                createdAt: this.state.createdAt.valueOf()//unix milleseconds
            })
        }
    }
    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input 
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input 
                        type="number"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <textarea 
                        placeholder="Add a note"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    ></textarea>
                    <SingleDatePicker
                        date={this.state.createdAt} // momentPropTypes.momentObj or null
                        onDateChange={date => this.setState({ createdAt: date })} // PropTypes.func.isRequired
                        focused={this.state.calendarFocused} // PropTypes.bool
                        onFocusChange={({ focused }) => this.setState({ calendarFocused: focused })}// PropTypes.func.isRequired
                        id="calendar-picker" // PropTypes.string.isRequired,
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <button type="submit">Add expense</button>
                </form>
            </div>
        )
    }
}