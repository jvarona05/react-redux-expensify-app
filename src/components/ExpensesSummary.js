import React from 'react';
import { connect } from 'react-redux'
import numeral from 'numeral'
import selectExpensesTotal from '../selectors/expenses-total'
import selectExpenses from '../selectors/expenses' 

export const ExpensesSummary = ({totalExpenses, totalAmount}) => {
    const expenseWord = totalExpenses === 1 ? 'expense' : 'expenses'
    const totalAmountFormated = numeral(totalAmount).format('$0,0.00')

    return (
        <div>
            <p>Viewing {totalExpenses} {expenseWord} totalling {totalAmountFormated}</p>
        </div>
    )
}

const mapStateToProps = (state) => {
    const visibledExpenses = selectExpenses(state.expenses, state.filters)

    return {
        totalAmount: selectExpensesTotal(visibledExpenses),
        totalExpenses: visibledExpenses.length
    }
}

export default connect(mapStateToProps)(ExpensesSummary)