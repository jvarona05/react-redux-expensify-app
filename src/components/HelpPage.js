import React from 'react';

const HelpPage = () => (<p>help</p>)

export default HelpPage