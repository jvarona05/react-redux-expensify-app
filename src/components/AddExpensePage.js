import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux'
import { addExpense } from '../actions/expenses'

export class AddExpensePage extends React.Component {
    onSubmit=(expense) => {
        this.props.addExpense(expense)
        this.props.history.push('/')
    }
    render() {
        return (
            <div>
                <ExpenseForm onSubmit={this.onSubmit} />
            </div>
        )
    }
}

//mapeamos el dispatch para poder testearlo de forma mas facil
const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense))
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage)