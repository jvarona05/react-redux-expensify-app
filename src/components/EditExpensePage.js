import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux'
import { editExpense, removeExpense } from '../actions/expenses'

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense)
        this.props.history.push('/')
    }
    onRemove = () => {
        this.props.removeExpense(this.props.expense.id)
        this.props.history.push('/')
    }
    render() {
        return (
            <div>
                <ExpenseForm 
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onRemove}>
                    remove
                </button>
            </div>
        )
    }
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find(({id}) => id === props.match.params.id)
    }
}

const mapDispatchToProps = (dispatch, props) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: (id) => dispatch(removeExpense(id)),
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)