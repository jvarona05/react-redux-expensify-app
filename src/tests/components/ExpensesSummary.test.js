import React from 'react'
import { shallow } from 'enzyme'
import { ExpensesSummary } from '../../components/ExpensesSummary'

test('should correctly render ExpensesSummary with 1 expense', () => {
    const wrapper = shallow(<ExpensesSummary totalExpenses={1} totalAmount={255} />)

    expect(wrapper).toMatchSnapshot()
})

test('should correctly render ExpensesSummary with multiples expenses', () => {
    const wrapper = shallow(<ExpensesSummary totalExpenses={4} totalAmount={4500} />)

    expect(wrapper).toMatchSnapshot()
})