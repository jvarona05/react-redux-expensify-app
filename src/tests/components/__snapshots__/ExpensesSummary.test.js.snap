// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSummary with 1 expense 1`] = `
<div>
  <p>
    Viewing 
    1
     
    expense
     totalling 
    $255.00
  </p>
</div>
`;

exports[`should correctly render ExpensesSummary with multiples expenses 1`] = `
<div>
  <p>
    Viewing 
    4
     
    expenses
     totalling 
    $4,500.00
  </p>
</div>
`;
